### PR TITLE
[MRESOLVER-241] Resolver checksum calculation should be driven by layout

### PR DESCRIPTION
**Pull request by cstamas**
Refactor checksum calculator and validator to rely on layout
provided checksums (layout always had these), and not by the presence
(or absence) or checksum locations.

Checksum locations should drive only the checksum provision from
remote, and nothing else.

### CODE REVIEW

**Review by slawekjaranowski — COMMENTED** [review]
I don't know this component ... but after read not found any obvious issues

**Comment by slawekjaranowski** [comment]
small request - try to align arguments in method calls like in:
https://github.com/apache/maven-site/pull/291

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java [hunk 1]
```diff
@@ -225,32 +239,9 @@ public URI getLocation( Metadata metadata, boolean upload )
...
-
-    private static class Maven2RepositoryLayoutEx
-            extends Maven2RepositoryLayout
-    {
-
-        protected Maven2RepositoryLayoutEx( List<ChecksumAlgorithmFactory> checksumAlgorithms )
+        private boolean isChecksum( String extension )
         {
-            super( checksumAlgorithms );
+            return checksumAlgorithms.stream().anyMatch( a -> extension.endsWith( "." + a.getFileExtension() ) );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:244]: Why `endsWith()` and not `equals()`?

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:244]: as extension of POM signature is `pom.asc`, also if you look at old code, it was doing same thing for same reason.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ */
+@Singleton
+@Named
+public class DefaultArtifactChecksumFilterFactory
+        implements ArtifactChecksumFilterFactory
+{
+    public static final String CONFIG_PROP_OMIT_CHECKSUMS_FOR_EXTENSIONS =
+            "aether.checksums.omitChecksumsForExtensions";
+
+    private static final String DEFAULT_OMIT_CHECKSUMS_FOR_EXTENSIONS = ".asc";
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: Since `org.eclipse.aether.artifact.Artifact.getExtension()` works without leading period, I would expect the same here.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            this.extensionsWithoutChecksums = requireNonNull( extensionsWithoutChecksums );
+        }
+
+        @Override
+        public boolean omitChecksumsFor( Artifact artifact )
+        {
+            String artifactExtension = artifact.getExtension(); // ie. pom.asc
+            for ( String extensionWithoutChecksums : extensionsWithoutChecksums )
+            {
+                if ( artifactExtension.endsWith( extensionWithoutChecksums ) )
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:82]: The period should be added for testing here

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java [hunk 4]
```diff
@@ -206,6 +216,10 @@ public URI getLocation( Metadata metadata, boolean upload )
         @Override
         public List<ChecksumLocation> getChecksumLocations( Artifact artifact, boolean upload, URI location )
         {
+            if ( artifactChecksumFilter.omitChecksumsFor( artifact ) || isChecksum( artifact.getExtension() ) )
```
[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:219]: artifact.getExtension returns values like "pom", "pom.sha1", "pom.asc" etc

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java (hunk 2 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: The signature artifact.getExtension returns "jar.asc", and we used `endsWith(.asc)` for same reason. Basically I did not change what happens in old code, just shuffled it around.

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: Since you have have basically replaced the old property I would strive for a cleanup and consistency with `#getExtension()`.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java (hunk 1 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:244]: True, the `a` provides only `md5`, etc. and as you noticed it is `XXX.md5`. This makes sense.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java (hunk 3 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:82]: See doco and default value: `.asc`, period is there. But right, if I introduce new packaging type "foo" (so extension is ".foo"), I'd be unable to make it filtered out...

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:82]: I think is ok: in my case of extension foo, I'd use value ".asc,foo" (so no leading dot before foo), while for sub-artifact like signature, value to be used is `.asc` (w/ leading dot).

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:82]: I don't fully understand your second comment. Does this mean that:
* Artifact Ext `X` should not be subject for checksum (without dot)
* Subartifact Ext `X.asc` needs the dot be differentiate between artifact and subartifact?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java (hunk 2 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: Having or not having leading period is important:
| inout  | `artifact-1.0.foo` | `artifact-1.0.jar.foo` |
|-------|------------------|---------------------|
|`.foo`  | not matches | matches |
|`foo`| metches | matches |

Some extensions (like PGP signature `asc`) may be only subartifact, so extension is actually `jar.asc` or `pom.asc`. Still, if we'd want to generalize this, we may do:
* if `extensionWithoutChecksums` starts with period => check with `endsWith`
* if `extensionWithoutChecksums` not starts with period => check with `equals`

WDYT?

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: OK, then I did understand your intention correctly. Yes, you proposal makes sense and you clearly convey the difference. Please also document it in the config.md file.

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: But let's not complicate, let's agree that:
* this PR w/o any custom config (in this area) behaves as master resolver w/o any custom config (jn this area)
* this PR w/ `aether.checksums.omitChecksumsForExtensions=""` (empty string) behaves the same as master does with `aether.checksums.forSignatures=true`
* for start, let's document these are for sub-artifacts only (signatures) and MUST use leading period.

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:52]: Agreed.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java (hunk 3 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/checksum/DefaultArtifactChecksumFilterFactory.java:82]: See https://github.com/apache/maven-resolver/pull/154#discussion_r825358528

**Comment by michael-o** [comment]
@jvanzyl Should we add `.sig` right away to the default list?

**Comment by jvanzyl** [comment]
If you're referring to Sigstore integration, it's up for discussion what the best naming conventions will be. I have x509, SSH, and PGP signing working. So probably want to keep `.asc` as not to interfere with existing PGP sigs, but for x509 maybe `.sig` and for SSH `.sshsig`. Not sure yet really.

**Comment by michael-o** [comment]
> If you're referring to Sigstore integration, it's up for discussion what the best naming conventions will be. I have x509, SSH, and PGP signing working. So probably want to keep `.asc` as not to interfere with existing PGP sigs, but for x509 maybe `.sig` and for SSH `.sshsig`. Not sure yet really.

OK, let's move this to a separate discussion when we have somthing usable.

File: maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.eclipse.aether.artifact.Artifact;
+
+/**
+ * Filter that is able to tell does artifact have expected checksums or not. Most notably, artifacts like
+ * different kind of signatures will not have checksums.
+ *
+ * @since 1.8.0
+ */
+public interface ArtifactChecksumFilter
```
Comment (cstamas) [inline maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java:30]: We should clarify here: this is ONLY FOR REMOTE_EXTERNAL artifacts. So if this class returns `true`, on deploy checksums will NOT be calculated and deployed, BUT on retrieve, the stream checksums ARE calculated (to satisfy PROVIDED or REMOTE_INCLUDED), but REMOTE_EXTERNAL ones will NOT be fetched (and hence will not be validated).

Comment (cstamas) [reply maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java:30]: fun fact: Central for example (but same stands for Nx2 or Artifactory) does sent checksum header for ANY payload, hence, even checksum for checksum, and resolver in this PR DOES USE IT to validate payload.

Comment (cstamas) [reply maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java:30]: Another minor thing: am not quite ok with class name `ArtifactChecksumFilter`. This class basically tells does passed in Artifact should have checksums or not... and it does not "filter" per se...

Comment (michael-o) [reply maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java:30]: `ArtifactChecksumDecider`?

Comment (cstamas) [reply maven-resolver-spi/src/main/java/org/eclipse/aether/spi/connector/checksum/ArtifactChecksumFilter.java:30]: hm, this needs more refactoring: I just realized, this is internal matter of Maven2RepositoryLayoutFactory. Still, IMO we do want to have it extensible, but as is in this PR, it seems misleading, as it's factory is really neglecting passed in repository completely, while all this applies to maven2 layout only. Also, I don't think a generic factory is right here, hence, I dislike to copy+paste same logic as in maven2 layout factory... Finally, I'd make this part of layout, not checksum package, maybe even just as a method on RepositoryLayout (to expose it) but it would be used internally by maven2 layout to decide...

**Comment by cstamas** [comment]
@michael-o review last commits pls, much simpler

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java [hunk 6]
```diff
@@ -118,9 +123,16 @@ public RepositoryLayout newInstance( RepositorySystemSession session, RemoteRepo
...
         }
 
-        return forSignature
-                ? new Maven2RepositoryLayout( checksumsAlgorithms )
-                : new Maven2RepositoryLayoutEx( checksumsAlgorithms );
+        // ensure uniqueness of (potentially user set) extension list
+        Set<String> omitChecksumsForExtensions = Arrays.stream( ConfigUtils.getString(
+                        session, DEFAULT_OMIT_CHECKSUMS_FOR_EXTENSIONS, CONFIG_PROP_OMIT_CHECKSUMS_FOR_EXTENSIONS )
+                .split( "," )
+        ).filter( s -> s != null && !s.trim().isEmpty() ).collect( Collectors.toSet() );
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:130]: Note: We don't apply `s != null && !s.trim().isEmpty()` to `CONFIG_PROP_CHECKSUMS_ALGORITHMS`. It should be consistent both or nothing.

**Review by michael-o — CHANGES_REQUESTED** [review]
What happended to `foo´,`.asc` dot and dotless?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java (hunk 6 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:129]: applied to both

**Comment by michael-o** [comment]
> What happended to `foo`,`.asc` dot and dotless?

Only this one is left. I will put this into testing meanwhile.

**Comment by cstamas** [comment]
IMO, as we still test with `endsWith`, dotless and dotted is important, but for simplicity sake, we'd maybe enforce configuration to use only dotted (dot prefixed) input. OTOH, I'd not apply anything like auto-adding dots, as that may hinder our later efforts to do something else (as we see that having or not having dot as prefix does matters, but we still don't know exactly what and how to use that, so better not allow users to be lazy that may hit us back later).

Simply put: change factory to enforce leading dots, and update doco that list must have extensions with leading dots.

Rationale:
* artifact.getExtension returns dotless extension (so not starting with dot), but sub-artifacts may have dots within (jar.asc)
* enforcing that users use dot-prefixed strings means we are applying this ONLY to sub-artifacts, that again, makes sense. I don't think we want to have top level artifacts that does NOT have checksums.

**Comment by michael-o** [comment]
Going through now...

**Review by michael-o — APPROVED** [review]
PR looks good now for me. Please squash, clean up commit message and merge.

**Comment by jira-importer** [comment]
Resolve #1366
